util/pingpong.c

```diff
@@ -0,0 +1,2229 @@
+
+	return str;
+}
+
+char *cnt_str(char str[PP_STR_LEN], long long cnt)
+{
+	if (cnt >= 1000000000)
+		snprintf(str, PP_STR_LEN, "%lldb", cnt / 1000000000);
+	else if (cnt >= 1000000)
+		snprintf(str, PP_STR_LEN, "%lldm", cnt / 1000000);
+	else if (cnt >= 1000)
+		snprintf(str, PP_STR_LEN, "%lldk", cnt / 1000);
+	else
+		snprintf(str, PP_STR_LEN, "%lld", cnt);
+
+	return str;
+}
+
+int64_t get_elapsed(const struct timespec *b, const struct timespec *a,
+		    enum precision p)
+{
+	int64_t elapsed;
+
+	elapsed = difftime(a->tv_sec, b->tv_sec) * 1000 * 1000 * 1000;
+	elapsed += a->tv_nsec - b->tv_nsec;
+	return elapsed / p;
+}
+
+void show_perf(char *name, int tsize, int sent, int acked, struct timespec *start,
+		struct timespec *end, int xfers_per_iter)
+{
+	static int header = 1;
+	char str[PP_STR_LEN];
+	int64_t elapsed = get_elapsed(start, end, MICRO);
+	long long bytes = (long long) sent * tsize * xfers_per_iter;
+	float usec_per_xfer;
+
+	if (sent == 0)
+		return;
+
+	if (name) {
+		if (header) {
+			printf("%-50s%-8s%-8s%-9s%-8s%8s %10s%13s%13s\n",
+					"name", "bytes", "#sent", "#ack",
+					"total", "time", "MB/sec",
+					"usec/xfer", "Mxfers/sec");
+			header = 0;
+		}
+
+		printf("%-50s", name);
+	} else {
+		if (header) {
+			printf("%-8s%-8s%-9s%-8s%8s %10s%13s%13s\n",
+					"bytes", "#sent", "#ack", "total",
+					"time", "MB/sec", "usec/xfer",
+					"Mxfers/sec");
+			header = 0;
+		}
+	}
+
+	printf("%-8s", size_str(str, tsize));
+
+	printf("%-8s", cnt_str(str, sent));
+
+	if (sent == acked) {
+		printf("=%-8s", cnt_str(str, acked));
+	} else if (sent < acked) {
+		printf("-%-8s", cnt_str(str, abs(acked - sent)));
+	} else {
+		printf("+%-8s", cnt_str(str, abs(acked - sent)));
+	}
+
+	printf("%-8s", size_str(str, bytes));
+
+	usec_per_xfer = ((float)elapsed / sent / xfers_per_iter);
+	printf("%8.2fs%10.2f%11.2f%11.2f\n",
+		elapsed / 1000000.0, bytes / (1.0 * elapsed),
+		usec_per_xfer, 1.0/usec_per_xfer);
+}
+
+/*******************************************************************************************/
+/*                                      Data Messaging                                     */
+/*******************************************************************************************/
+
+int pp_cq_readerr(struct fid_cq *cq)
+{
+	struct fi_cq_err_entry cq_err;
+	int ret;
+
+	ret = fi_cq_readerr(cq, &cq_err, 0);
+	if (ret < 0) {
+		PP_PRINTERR("fi_cq_readerr", ret);
+	} else {
+		PP_ERR("cq_readerr: %s", fi_cq_strerror(cq, cq_err.prov_errno,
+					cq_err.err_data, NULL, 0));
+		ret = -cq_err.err;
+	}
+	return ret;
+}
+
+/*
+ * fi_cq_err_entry can be cast to any CQ entry format.
+ */
+static int pp_spin_for_comp(struct fid_cq *cq, uint64_t *cur,
+			    uint64_t total, int timeout)
+{
+	struct fi_cq_err_entry comp;
+	struct timespec a, b;
+	int ret;
+
+	if (timeout >= 0)
+		clock_gettime(CLOCK_MONOTONIC, &a);
+
+	while (total - *cur > 0) {
+		ret = fi_cq_read(cq, &comp, 1);
```

Comment:
Is there a performance difference between cq_read() vs. cq_sread()?  If there is a difference, it would be better to use the higher-performing one here.